tasks/windows.ps1

```diff
@@ -0,0 +1,41 @@
+[CmdletBinding()]
+Param(
+  [Parameter(Mandatory = $True)] [String] $master,
+  [Parameter(Mandatory = $False)] [String] $cacert_content,
+  [Parameter(Mandatory = $False)] [String] $certname,
+  [Parameter(Mandatory = $False)] [String] $dns_alt_names
+)
+
+Function Validate-Parameter {
+  Param ($param)
+  if ($param -contains "'") {
+    Throw 'Single-quote is not allowed in arguments'
+  }
+}
+
+Validate-Parameter($certname)
+Validate-Parameter($alt_names)
+
+if ($certname.Length -eq 0) {
+  $certname_arg = "agent:certname='$certname' "
+}
+else {
+  $fqdn = [System.Net.Dns]::GetHostByName(($env:computerName)).Hostname 
+  $certname_arg = "agent:certname='$fqdn' "
+}
+if ($dns_alt_names.Length -eq 0) {
+  $alt_names_arg = "agent:dns_alt_names='$alt_names' "
+}
+else {
+  $alt_names_arg = ""
+}
+if ($cacert_content) {
```

Comment:
This is going to be a bit tricky... creating the files / folders here has the potential to lay down permissions in a way that either creates a security issue by being too permissive OR prevents Puppet from accessing the files. This should probably be written *after* the agent is installed and those directories have been created... but not sure if that puts us into a chicken and egg situation.

Reply:
Same with the puppet.conf file